IOS-6054 Fix send tx error model

skibinalexander opened the pull request:
Данные решение обсуждали с @tureck1y. Проблема была следующего характера: пользователь при не успешной отправке транзакции формировал отчет, и там должен захватываться хост, при этом он захватывается не тот на котором было исключение, а уже переключенный на 0 порядок. Также обсудили с Сашей необходимость изменения протокола.

Мы с Сашей обсудили что захватывать хост в WalletManager, получалось немного грязно, и очень легко было забыть, поэтому я вынес ошибку в мультинетворк провайдеры и делаю захват хоста там.

skibinalexander committed:
IOS-6054 Fix send tx error model

skibinalexander commented on BlockchainSdk/Blockchains/Hedera/HederaWalletManager.swift:
```diff
@@ -377,14 +377,22 @@ extension HederaWalletManager: WalletManager {
-            }
-        )
-        .map(\.1)
+        .mapSendError()
```
@m3g0byt3 надо нам с тобой тут посмотреть

skibinalexander commented on BlockchainSdk/Blockchains/VeChain/VeChainWalletManager.swift:
```diff
@@ -259,16 +259,21 @@ extension VeChainWalletManager: WalletManager {
-                }
-            )
-            .map(\.1)
+            .mapSendError()
```
@m3g0byt3 надо нам с тобой тут посмотреть

skibinalexander commented:
Тесты поправлю

Andoran90 commented on BlockchainSdk/Blockchains/XRP/XRPWalletManager.swift:
```diff
@@ -105,19 +105,21 @@ extension XRPWalletManager: TransactionSender {
+                walletManager.networkService.send(blob: hash)
                     .tryMap{[weak self] response in
-                        guard let self = self else { throw WalletError.empty }
+                        guard let self else { throw SendTxError(error: WalletError.empty) }
```
тут тоже лишнее оборачивание в `SendTxError`

skibinalexander committed:
IOS-6054 Fix review comments

skibinalexander committed:
IOS-6054 Fix tests

skibinalexander committed:
IOS-6054 Fix after discuss

m3g0byt3 commented on BlockchainSdk/Blockchains/Algorand/AlgorandWalletManager.swift:
```diff
@@ -154,7 +154,10 @@ extension AlgorandWalletManager {
-                return walletManager.networkService.sendTransaction(data: transactionData)
+                return walletManager.networkService
+                    .sendTransaction(data: transactionData)
+                    .mapError { SendTxError(error: $0, tx: transactionData.hexString.lowercased()) }
```
nit: Разброс у нас какой-то небольшой с экстеншенами.

В BSDK вроде уже есть `hexadecimal`, который в lower case - то есть для него нет необходимости делать `lowercased()`

А `hexString` из TSDK действительно в upper case

Какой экстеншн трушнее?

skibinalexander replied:
> nit: Разброс у нас какой-то небольшой с экстеншенами.
> 
> В BSDK вроде уже есть `hexadecimal`, который в lower case - то есть для него нет необходимости делать `lowercased()`
> 
> А `hexString` из TSDK действительно в upper case
> 
> Какой экстеншн трушнее?

Так вроде по последним ревью @tureck1y командовал hexString

tureck1y replied:
hexString это наш сдкшный, лучше его + lowercase если надо

tureck1y replied:
Чтобы не сменилось вдруг поведение из-за выкапывания/добавления какой-то либы

m3g0byt3 commented on BlockchainSdk/Blockchains/Bitcoin/BitcoinWalletManager.swift:
```diff
@@ -90,13 +90,18 @@ class BitcoinWalletManager: BaseManager, WalletManager, DustRestrictable {
             }
             .flatMap {[weak self] tx -> AnyPublisher<TransactionSendResult, Error> in
-                guard let self = self else { return .emptyFail }
+                guard let self else { return .anyFail(error: WalletError.empty) }
```
nit: а что не так с `emptyFail` ?
вроде тоже самое под капотом

skibinalexander replied:
> nit: а что не так с `emptyFail` ? вроде тоже самое под капотом
Ага , спасибо, рудимент перебора вариантов

m3g0byt3 commented on BlockchainSdk/Blockchains/Hedera/HederaWalletManager.swift:
```diff
@@ -377,7 +377,9 @@ extension HederaWalletManager: WalletManager {
             return walletManager
                 .networkService
                 .send(transaction: compiledTransaction)
+                .eraseToAnyPublisher()
```
nit: разве нужно тут? Я про `eraseToAnyPublisher`, кажется и без него типы выводятся

m3g0byt3 commented on BlockchainSdk/Blockchains/Hedera/HederaWalletManager.swift:
```diff
@@ -377,7 +377,9 @@ extension HederaWalletManager: WalletManager {
             return walletManager
```
А тут кажется нет оборачивания в `SendError`?

m3g0byt3 commented on BlockchainSdk/Blockchains/NEAR/NEARWalletManager.swift:
```diff
@@ -254,9 +254,15 @@ extension NEARWalletManager: WalletManager {
+                    .mapError { error in
+                        SendTxError(error: error, tx: rawTransactionData.hexString.lowercased())
+                    }
+                    .eraseToAnyPublisher()
```
Тоже не очень понимаю зачем, ниже по цепочке вроде и так будет обернут
А типы и без `AnyPublisher`, кажется выводятся

m3g0byt3 commented on BlockchainSdk/Blockchains/Solana/SolanaWalletManager.swift:
```diff
@@ -76,6 +76,7 @@ extension SolanaWalletManager: TransactionSender {
                 self.wallet.addPendingTransaction(record)
                 return TransactionSendResult(hash: hash)
```
А тут кажется нет оборачивания в `SendError`?

m3g0byt3 commented on BlockchainSdk/Blockchains/VeChain/VeChainWalletManager.swift:
```diff
@@ -259,9 +259,14 @@ extension VeChainWalletManager: WalletManager {
+                    .networkService
+                    .send(transaction: rawTransactionData)
+                    .mapError { SendTxError(error: $0, tx: rawTransactionData.hexString.lowercased()) }
+                    .eraseToAnyPublisher()
```
nit: разве нужно тут? Я про `eraseToAnyPublisher`, кажется и без него типы выводятся

m3g0byt3 commented on BlockchainSdk/Common/API/MultiNetworkProvider.swift:
```diff
@@ -92,3 +94,30 @@ extension MultiNetworkProvider {
+        (networkError as? MoyaError)?.helpAnchor ?? defaultMoyaError.helpAnchor
+    }
+    
+    var defaultMoyaError: MoyaError {
+        .underlying(networkError, nil)
+    }
```
Лучше сделать приватным, если это будет торчать наружу - то непонятно зачем и как это юзать

m3g0byt3 commented on BlockchainSdk/Common/API/MultiNetworkProvider.swift:
```diff
@@ -67,7 +67,9 @@ extension MultiNetworkProvider {
                 }
                 
-                return .anyFail(error: error)
+                // Need captured currentHost, to be able to get hosting after switching.
```
Этот коммент валиден скорее к 38 строчке, где собственно и сохраняется текущий хост

skibinalexander replied:
> Этот коммент валиден скорее к 38 строчке, где собственно и сохраняется текущий хост

Да там как раз понятно, здесь валютно для того чтобы понять , почему вообще ошибку оборачиваем

m3g0byt3 commented on BlockchainSdk/Common/TransactionSendResult.swift:
```diff
@@ -15,3 +15,34 @@ public struct TransactionSendResult {
+    // MARK: - Init
+    
+    init(error: Error, tx: String? = nil) {
+        // Need be use ....
+        if let sendError = error as? Self {
+            self = sendError
+        } else if let providerError = error as? MultiNetworkProviderError {
+            self.error = providerError.networkError
+            self.tx = tx
+            self.lastRetryHost = providerError.lastRetryHost
+        } else {
+            self.error = error
+            self.tx = tx
+            self.lastRetryHost = nil
+        }
```
А эту логику нe уместнее держать в хелпере `mapSendError`?
Не обсуждали это с Сашей?

А то как будто эта структура слишком много знает

skibinalexander replied:
Давай попробуем, ага помню обсуждали

m3g0byt3 reviewed (commented):
1. Сейчас в работе JoyStream, для него не понадобится доработок?
2. Может добавить новый негативный юнит тест, который:
- Проходится по `Blockchain.allMainnetCases` и создает воллет менеджеры
- Отправляет заведомо неудачную транзу (тут я честно говоря не знаю, можно ли универсально сделать всегда неуспешную транзу или для каждой сети надо задавать свои параметры для этого)
- Проверяет, что такая транзакция зафейлилась и то, что у результирующего `SendTxError` есть `raw tx`

Иначе мне кажется разъедется это в будущем, будут забывать положить `raw tx` в `SendTxError` в новых сетях

skibinalexander commented:
> 1. Сейчас в работе JoyStream, для него не понадобится доработок?
> 2. Может добавить новый негативный юнит тест, который:
> 
> * Проходится по `Blockchain.allMainnetCases` и создает воллет менеджеры
> * Отправляет заведомо неудачную транзу (тут я честно говоря не знаю, можно ли универсально сделать всегда неуспешную транзу или для каждой сети надо задавать свои параметры для этого)
> * Проверяет, что такая транзакция зафейлилась и то, что у результирующего `SendTxError` есть `raw tx`
> 
> Иначе мне кажется разъедется это в будущем, будут забывать положить `raw tx` в `SendTxError` в новых сетях

Я не уверен, что это нужно делать в рамках текущей задачи, понятно, что можно забыть, но в целом на ревью посмотреть. Просто сейчас городить огород с Dummy фабрикой менеджеров выглядит избыточно

skibinalexander commented:
> 1. Сейчас в работе JoyStream, для него не понадобится доработок?
> 2. Может добавить новый негативный юнит тест, который:
> 
> * Проходится по `Blockchain.allMainnetCases` и создает воллет менеджеры
> * Отправляет заведомо неудачную транзу (тут я честно говоря не знаю, можно ли универсально сделать всегда неуспешную транзу или для каждой сети надо задавать свои параметры для этого)
> * Проверяет, что такая транзакция зафейлилась и то, что у результирующего `SendTxError` есть `raw tx`
> 
> Иначе мне кажется разъедется это в будущем, будут забывать положить `raw tx` в `SendTxError` в новых сетях

Еще же протокол в принципе поменяется, заставит обратить внимание

skibinalexander committed:
IOS-6054 Fix review comments

skibinalexander commented on BlockchainSdk.xcworkspace/xcshareddata/swiftpm/Package.resolved:
```diff
@@ -1,5 +1,5 @@
 {
-  "originHash" : "09ecb99db3e580b585d5d932184af81c65d50399b08821d46e7de1ac03c0fdd2",
+  "originHash" : "a0738665bda10cbdb05da546422ce445e2fdf9ace530c7941d4cba18d71a3d25",
```
Тут в диф что-то попало, уберу

skibinalexander committed:
IOS-6054 Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

# Conflicts:
#	BlockchainSdkExample/BlockchainSdkExampleViewModel.swift

skibinalexander committed:
IOS-6054 Fix package.resolved

skibinalexander committed:
IOS-6054 Revert changes

Andoran90 commented on BlockchainSdk/Extensions/Publisher+.swift:
```diff
@@ -85,6 +90,26 @@ extension Publisher {
+extension Publisher where Failure == Error {
+    func mapSendError(tx: String? = nil) -> Publishers.MapError<Self, Error> {
+        mapError { error in
+            if let sendError = error as? SendTxError {
+                return sendError
+            } else if let providerError = error as? MultiNetworkProviderError {
+                return SendTxError(error: providerError.networkError, tx: tx, lastRetryHost: providerError.lastRetryHost)
+            } else {
+                return SendTxError(error: error, tx: tx)
+            }
```
как вариант можно переписать:
``` swift
switch error {
case let sendError as SendTxError:
    return sendError
default:
    return SendTxError(error: error, tx: tx)
```

Andoran90 commented on BlockchainSdk/Extensions/Publisher+.swift:
```diff
@@ -85,6 +90,26 @@ extension Publisher {
+        }
+    }
+    
+    func eraseSendError() -> Publishers.MapError<Self, SendTxError> {
+        mapError { error in
+            SendTxError(error: error)
+        }
+    }
```
думаю не очень хорошая идея что у нас есть 2 варианта: `mapSendError` и `eraseSendError`, может получиться так, что сделаем матрешку `SendTxError`. Думаю эта функция вообще лишняя, и достаточно `mapSendError()`, которая может быть и без параметров и через неё гонять всю логику.

Да и вообще когда будешь писать код будет возникать вопрос - какую из функций использовать, чтобы правильно обернуть в ошибку и пока реализацию не посмотришь, не поймешь что лучше использовать `mapSendError`))

skibinalexander replied:
> думаю не очень хорошая идея что у нас есть 2 варианта: `mapSendError` и `eraseSendError`, может получиться так, что сделаем матрешку `SendTxError`. Думаю эта функция вообще лишняя, и достаточно `mapSendError()`, которая может быть и без параметров и через неё гонять всю логику.
> 
> Да и вообще когда будешь писать код будет возникать вопрос - какую из функций использовать, чтобы правильно обернуть в ошибку и пока реализацию не посмотришь, не поймешь что лучше использовать `mapSendError`))

Ну тогда ревертну , как изначально было написано в инициализаторе, Андрюхе просто это не очень понравилось

Andoran90 replied:
нене, погодь, полностью ревертить не надо

Andoran90 replied:
мне не нравится что есть 2 функции, надо оставить только `mapSendError` и через него все гнать

Andoran90 commented on BlockchainSdk/Blockchains/Tezos/TezosWalletManager.swift:
```diff
@@ -85,23 +85,25 @@ extension TezosWalletManager: TransactionSender {
+                        return TransactionSendResult(hash: rawTransaction)
                     }
-                    .mapError { SendTxError(error: $0, tx: hash) }
+                    .mapSendError(tx: rawTransaction)
                     .eraseToAnyPublisher()
             }
+            .eraseSendError()
```
вот тут как раз получится матрешка, если при попытке отправить транзакцию будет ошибка, мы в начале обернем сетевую ошибку в `SendTxError`, а потом её же положим в `SendTxError`

skibinalexander committed:
IOS-6054 Fix review comments

skibinalexander committed:
IOS-6054 Fix review comments

skibinalexander committed:
IOS-6054 Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

# Conflicts:
#	BlockchainSdk/Blockchains/Binance/BinanceWalletManager.swift

skibinalexander commented:
Обсудили с @Andoran90 добавил фабрику для того чтобы логика генерации ошибка была в одном месте, и не получилось вложенности SendTxError на SendTxError. Привести к одному методу не получилось, потому что один возвращает any Error другой по сути финальное в цепочке SendTxError

tureck1y commented on BlockchainSdk/Blockchains/Algorand/AlgorandWalletManager.swift:
```diff
@@ -154,7 +154,10 @@ extension AlgorandWalletManager {
             }
             .withWeakCaptureOf(self)
             .flatMap { walletManager, transactionData -> AnyPublisher<String, Error> in
-                return walletManager.networkService.sendTransaction(data: transactionData)
+                return walletManager.networkService
```
а если написать тут -> AnyPublisher<String, SendError>, то может не понадобится делать erase ниже? В остальных менеджерах то же самое в теории

skibinalexander replied:
> а если написать тут -> AnyPublisher<String, SendError>, то может не понадобится делать erase ниже? В остальных менеджерах то же самое в теории

Не для всех менеджеров, где-то есть глубина в слои, и туда тоже лезут эти изменения, и не везде выходит унификация .... и получается куча изменений. Где то конфликты внутри flatMap идут.

tureck1y commented on BlockchainSdk/Blockchains/Binance/BinanceWalletManager.swift:
```diff
@@ -98,7 +101,7 @@ extension BinanceWalletManager: TransactionFeeProvider {
                 guard let self = self else { throw WalletError.empty }
                 
-                guard let feeValue = Decimal(feeString) else {
+                guard let feeValue = Decimal(string: feeString) else {
```
а для чего это изменение?

skibinalexander replied:
> а для чего это изменение?

Да пробегался по менеджерам, в целом изменения залезли некоторые, решил тут не проходить мимо, и warning поправить. с @m3g0byt3 как раз обсуждали в других МР что надо бы переехать на этот инициализатор

tureck1y replied:
давай лучше везде централизованно поменяем отдельной задачей, потенциально может что-то сломать

skibinalexander replied:
https://tangem.atlassian.net/browse/IOS-6870

skibinalexander committed:
Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

Andoran90 commented on BlockchainSdk/Blockchains/Binance/BinanceWalletManager.swift:
```diff
@@ -78,8 +78,11 @@ extension BinanceWalletManager: TransactionSender {
                     return TransactionSendResult(hash: hash)
 
-                }.eraseToAnyPublisher() ?? .emptyFail
+                    }
+                    .mapSendError(tx: tx.encodeForSignature().hexString.lowercased())
+                    .eraseToAnyPublisher() ?? .emptyFail
```
отформатировать код надо, отступ неправильный)

Andoran90 commented on BlockchainSdk/Blockchains/Bitcoin/BitcoinWalletManager.swift:
```diff
@@ -90,13 +90,18 @@ class BitcoinWalletManager: BaseManager, WalletManager, DustRestrictable {
-                    txHashPublisher = self.networkService.push(transaction: tx)
+                    txHashPublisher = self.networkService
+                        .push(transaction: tx)
+                        .eraseToAnyPublisher()
```
а зачем туть это и на 104 строке? как будто лишнее, т.к. и там и там возвращаются `AnyPublisher<String, Error>`

Balashov152 commented on BlockchainSdk/Blockchains/Bitcoin/BitcoinWalletManager.swift:
```diff
@@ -166,7 +172,7 @@ extension BitcoinWalletManager: BitcoinTransactionFeeCalculator {
 @available(iOS 13.0, *)
 extension BitcoinWalletManager: TransactionSender {
-    func send(_ transaction: Transaction, signer: TransactionSigner) -> AnyPublisher<TransactionSendResult, Error> {
+    func send(_ transaction: Transaction, signer: TransactionSigner) -> AnyPublisher<TransactionSendResult, SendTxError> {
```
может тут вернуть 
```suggestion
    func send(_ transaction: Transaction, signer: TransactionSigner) -> some Publisher<TransactionSendResult, SendTxError> {
```

много `eraseToAnyPublisher` уйдет

skibinalexander replied:
@Balashov152 положил в беклог, Этот мр и так разросся

https://tangem.atlassian.net/browse/IOS-6865

skibinalexander committed:
IOS-6054 Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

# Conflicts:
#	BlockchainSdk/Blockchains/Stellar/StellarWalletManager.swift
#	BlockchainSdk/Blockchains/XRP/XRPWalletManager.swift

skibinalexander committed:
IOS-6054 Fix tracked tx hash

skibinalexander committed:
IOS-6054 Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

skibinalexander committed:
IOS-6054 Fix error after tests

skibinalexander commented:
@tureck1y потестировал через приложение работает корректно

skibinalexander commented:
Можно возвращать галочки

tureck1y reviewed (commented):
думаю надо лишние правки с конвертацией все же откатить и сделать отдельно

skibinalexander commented:
> думаю надо лишние правки с конвертацией все же откатить и сделать отдельно

Оке , сделаю

skibinalexander committed:
IOS-6054 Revert changes for Decimal(string)

skibinalexander commented:
@tureck1y отревертил по Decimal

skibinalexander commented:
Блин конфликты набежали, решаю

skibinalexander committed:
IOS-6054 Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

# Conflicts:
#	BlockchainSdk/Blockchains/Hedera/HederaWalletManager.swift

skibinalexander committed:
IOS-6054 Fix errors

skibinalexander committed:
Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

tureck1y reviewed (commented):
тесты лежат

skibinalexander committed:
IOS-6054 Merge branch 'develop' into bugfix/IOS-6054_make_send_tx_error_interface

skibinalexander committed:
IOS-6054 Revert changes for Package.resolved
